presale/models/presale_model.py

```diff
@@ -0,0 +1,66 @@
+from odoo import fields, models
+
+
+class PresaleModel(models.Model):
+    _name = "presale.order"
+    _description = "Presale order model"
+
+    def _defaultName(self):
+        # TODO: Avoid searching the whole table
+        number = 1
+        record_names = [record.name for record in self.env["presale.order"].search([])]
+        while ("PS " + str(number).zfill(3)) in record_names:
+            number += 1
+        return "PS " + str(number).zfill(3)
+
+    active = fields.Boolean(default=True)
+    name = fields.Char(default=_defaultName, required=True)
+    partner_id = fields.Many2one("res.partner", string="Customer")
+    state = fields.Selection(
+        default="new",
+        required=True,
+        selection=[("new", "Draft"), ("validated", "Validated"), ("cancelled", "Cancelled")],
+    )
+    order_line_ids = fields.One2many("presale.order.line", "presale_order_id")
+
+    def send_validation_email(self, presale_order):
+        # TODO: test with proper email
+        mail_template = {
+            "subject": presale_order.name,
+            "email_to": presale_order.partner_id.email,
+            "body_html": f"<h2>Your presale order {presale_order.name} has been validated</h2>",
+        }
+        msg_id = self.env["mail.mail"].create(mail_template)
+        self.env["mail.mail"].send([msg_id])
+
+    def action_validate_preorder(self):
+        for record in self:
+            if record.state == "new":
+                record.state = "validated"
+                sale_order = self.env["sale.order"].create(
+                    {
+                        "name": record.name,
+                        "partner_id": record.partner_id.id,
+                    }
+                )
+                sale_order.presale_order = record.id
+                for presale_order_line in record.order_line_ids:
```

Comment:
I think you should rewrite this so as to call create only once.

for example you could do 
```py
sale_order_lines = list()
for presale_order_line in record.order_line_ids:
             sale_order_lines.append({....})
self.env["sale.order.line"].create(sale_order_lines)
```